KAFKA-5179: Log connection termination during authentication

rajinisivaram opened the pull request.

rajinisivaram committed:
KAFKA-5179: Log connection termination during authentication for clients

ijuma commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -407,6 +407,16 @@ private void maybeRecordTimePerConnection(KafkaChannel channel, long startTimeNa
+            channel.prepare();
+        } catch (IOException e) {
+            log.warn("Connection with {} terminated during authentication. This may indicate " +
+                    "authentication failure due to invalid credentials", channel.socketDescription());
```
Nit: would it read better if we said "This may indicate that authentication failed due to invalid credentials"?

rajinisivaram commented:
@ijuma I have changed the code (quite a bit). I think we don't want to add warnings in the broker when clients disconnect. So I have moved the warning to `NetworkClient`. Have also added warnings for the connection refused case that you mentioned in the JIRA. The additional state in `KafkaChannel` will also be useful if we decide to go ahead with KIP-152 to have a more definitive authentication-failed state.

ijuma commented:
Thanks @rajinisivaram, will take a look soon.

ijuma commented:
An interesting question is how the additional state in `KafkaChannel` affects `ClusterConnectionStates`. The answer may be that it doesn't, but worth thinking about.

ijuma commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -521,8 +523,10 @@ private int select(long ms) throws IOException {
-        if (channel != null)
+        if (channel != null) {
+            channel.state(ChannelState.LOCAL_CLOSE);
             close(channel, false);
```
By passing `false` here, the channel is never added to the `disconnected` list and since we remove the channel from `channels`, it seems like this state transition is never observed. Is that correct?

rajinisivaram replied:
Yes, that is correct. I have added a comment. When NetworkClient closes connection (e.g. due to a timeout), it invokes `processDisconnection` after invoking `close`, but that doesn't rely on this state.

ijuma commented on clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java:
```diff
@@ -72,16 +74,28 @@ public void prepare() throws IOException {
-        return transportLayer.finishConnect();
+        boolean connected = transportLayer.finishConnect();
+        if (connected)
+            state = ready() ? ChannelState.READY : ChannelState.AUTHENTICATE;
```
Seems like we transition to AUTHENTICATE before the TLS handshake (if TLS is enabled). If TLS authentication fails, will we also get a warning then? I think that would be good, but just checking as previous proposals didn't cover TLS.

rajinisivaram replied:
Yes, the previous proposal was purely for SASL. But it made sense to do the warnings for both TLS and SASL since we can, so both TLS and SASL handshake failures result in the same warning with this PR.

ijuma commented on clients/src/test/java/org/apache/kafka/common/network/SelectorTest.java:
```diff
@@ -127,8 +127,10 @@ public void testConnectionRefused() throws Exception {
         selector.connect(node, new InetSocketAddress("localhost", nonListeningPort), BUFFER_SIZE, BUFFER_SIZE);
-        while (selector.disconnected().contains(node))
+        while (selector.disconnected().containsKey(node)) {
+            assertEquals(ChannelState.NOT_CONNECTED, selector.disconnected().get(node));
```
In the other PR, the an issue was raised that `disconnected` only gets populated if there are in flight requests. And if the broker is misconfigured, we never have in-flight request. How have we fixed that?

rajinisivaram replied:
If the broker is misconfigured, we see a `Connection Refused`. This does go through the disconnect notification code path. In `NetworkClient#processDisconnection`, the new warning added for `NOT_CONNECTED` handles this case.
```

[main] DEBUG Connection with localhost/127.0.0.1 disconnected (org.apache.kafka.common.network.Selector:394)
java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
	at org.apache.kafka.common.network.PlaintextTransportLayer.finishConnect(PlaintextTransportLayer.java:50)
	at org.apache.kafka.common.network.KafkaChannel.finishConnect(KafkaChannel.java:95)
	at org.apache.kafka.common.network.Selector.pollSelectionKeys(Selector.java:354)
	at org.apache.kafka.common.network.Selector.poll(Selector.java:321)
	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:359)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:230)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:206)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:197)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.awaitMetadataUpdate(ConsumerNetworkClient.java:132)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureCoordinatorReady(AbstractCoordinator.java:223)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureCoordinatorReady(AbstractCoordinator.java:200)
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.poll(ConsumerCoordinator.java:285)
	at org.apache.kafka.clients.consumer.KafkaConsumer.pollOnce(KafkaConsumer.java:1043)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1008)
	at kafka.consumer.NewShinyConsumer.<init>(BaseConsumer.scala:58)
	at kafka.tools.ConsoleConsumer$.run(ConsoleConsumer.scala:72)
	at kafka.tools.ConsoleConsumer$.main(ConsoleConsumer.scala:53)
	at kafka.tools.ConsoleConsumer.main(ConsoleConsumer.scala)
[main] DEBUG Node -1 disconnected. (org.apache.kafka.clients.NetworkClient:605)
[main] WARN Connection to node -1 could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient:496)
```

ijuma reviewed (commented):
Thanks for the PR. Left a few questions. Also, should we remove the logging code below?

```java
@Override
        public void handleDisconnection(String destination) {
            Cluster cluster = metadata.fetch();
            if (cluster.isBootstrapConfigured()) {
                int nodeId = Integer.parseInt(destination);
                Node node = cluster.nodeById(nodeId);
                if (node != null)
                    log.warn("Bootstrap broker {}:{} disconnected", node.host(), node.port());
            }

            metadataFetchInProgress = false;
        }
```

rajinisivaram committed:
KAFKA-5179: Add comment, fix log entry

rajinisivaram commented:
@ijuma Thank you for the review. 

I think it is better to keep `ClusterConnectionState` and `ChannelState` separate even though there is some overlap. It makes it easier to follow the code.

The warning in `handleDisconnection` is triggered when a connection to a bootstrap server is broken after the connection was established. The new warnings in this PR are for
1. misconfigured broker - connection could not be established
2. misconfigured security - authentication did not complete

I didn't want to remove an existing warning just in case it was useful. For broken connections, we do have debug logs. Do you think that is sufficient?

ijuma commented:
@rajinisivaram I see, the `handleDisconnection` warning is useful if there is a broker error that causes a disconnection (e.g. unsupported API version) although it seems less useful now that we do version negotiation. It may be worth adding a comment, but we can probably keep it for now.

rajinisivaram committed:
KAFKA-5179: Address review comment

rajinisivaram commented:
@ijuma Thank you, added a comment.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java:
```diff
@@ -72,16 +74,28 @@ public void prepare() throws IOException {
             transportLayer.handshake();
         if (transportLayer.ready() && !authenticator.complete())
             authenticator.authenticate();
+        if (transportLayer.ready() && authenticator.complete())
```
This can just be if (ready())?

rajinisivaram replied:
Yes, done.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -494,7 +495,8 @@ private void clear() {
         }
-        this.disconnected.addAll(this.failedSends);
+        for (String channel : this.failedSends)
+            this.disconnected.put(channel, ChannelState.FAILED_SEND);
```
Should we set the state of channel to FAILED_SEND too?

rajinisivaram replied:
Done.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/ChannelState.java:
```diff
@@ -0,0 +1,26 @@
+ */
+package org.apache.kafka.common.network;
+
+public enum ChannelState {
```
Could we add some comments that explain the meaning of each state and what the typical state transition flow looks like?

rajinisivaram replied:
@junrao Thank you for the review. I have added comments.

junrao reviewed (commented):
@rajinisivaram : Thanks for the patch. LGTM. Just a few minor comments.

rajinisivaram committed:
KAFKA-5179: Address review comments

junrao commented:
@rajinisivaram : Thanks for the patch. LGTM

rajinisivaram commented:
@junrao Thank you for the review. Merging to trunk.
